### PR TITLE
Enhancement/slepc

### PR DESCRIPTION
**Pull request by juan42**
Reassigning this pull request to #738, i'm sorry for the confusion but my branch wouldn't merge properly anymore.

This adds the functionality to install SLEPc with the firedrake-install script and includes a regression test based on the calculation of generalized physical eigenvalues for the Laplace operator.

### CODE REVIEW

File: tests/regression/test_slepc.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+def topetsc(A):
+    return A.M.handle
+
+
+def test_import():
+    from slepc4py import SLEPc
+
+
+def test_laplace_physical_ev(parallel=False):
+    from slepc4py import SLEPc
```
Comment (wence-) [inline tests/regression/test_slepc.py]: Since slepc is an optional dep, can you guard this test like so:

```python
try:
    from slepc4py import SLEPc
except ImportError:
    pytest.skip(msg="SLEPc unavailable, skipping eigenvalue test")
```

File: tests/regression/test_slepc.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+from firedrake import *
+from firedrake.petsc import PETSc
+
+
+def topetsc(A):
+    return A.M.handle
+
+
+def test_import():
+    from slepc4py import SLEPc
```
Comment (wence-) [inline tests/regression/test_slepc.py]: Also with this test (or just remove it).

File: scripts/firedrake-install [hunk 3]
```diff
@@ -491,7 +494,7 @@ def install(package):
             log.info("Installed %s from travis cache %s\n" % (package, travis_cache_dir))
             return
     # The following outrageous hack works around the fact that petsc cannot be installed in developer mode.
-    if args.developer and package not in ["petsc/", "petsc4py/"]:
+    if args.developer and package not in ["petsc/", "petsc4py/", "slepc/", "slepc4py/"]:
```
Comment (dham) [inline scripts/firedrake-install]: Is it really the case that developer mode is also broken for slepc?

File: scripts/firedrake-install [hunk 4]
```diff
@@ -482,7 +485,7 @@ if travis:
 
 def install(package):
     log.info("Installing %s" % package)
-    if travis and travis_cache_dir and package in ["petsc/", "petsc4py/"]:
+    if travis and travis_cache_dir and package in ["petsc/", "petsc4py/", "slepc/", "slepc4py/"]:
```
Comment (dham) [inline scripts/firedrake-install]: I don't think you mean this. This is the magic which causes petsc to not get rebuilt on travis. This is only needed because building petsc takes so long.

File: scripts/firedrake-install [hunk 5]
```diff
@@ -111,6 +112,8 @@ else:
                         help="Rebuild all packages even if no new version is available. Usually petsc and petsc4py are only rebuilt if they change. All other packages are always rebuilt.")
     parser.add_argument("--adjoint", action='store_true', dest="install_adjoint",
                         help="Also install firedrake-adjoint.")
+    parser.add_argument("--slepc", action="store_true",
+                        help="Install SLEPc along with PETSc")
```
Comment (dham) [inline scripts/firedrake-install]: This is not the right way to do this. You need to copy what happens for --adjoint

File: scripts/firedrake-install (hunk 3 above)
Comment (juan42) [inline scripts/firedrake-install]: The install script fails with the -e option.

File: tests/regression/test_slepc.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    A = topetsc(assemble(a, bcs=[bc]))
+    M = topetsc(assemble(m, bcs=[bc]))
+
+    # This shifts the "1.0" Eigenvalues out of the spectrum
+    # of interest (which is around 0.0 in this case).
+    vals = np.repeat(1E8, len(bc.nodes))
+    A.setValuesLocalRCV(bc.nodes.reshape(-1, 1),
+                        bc.nodes.reshape(-1, 1),
+                        vals.reshape(-1, 1))
+    A.assemble()
```
Comment (wence-) [inline tests/regression/test_slepc.py]: I think if you didn't want to do this messing with the operator, you could instead assembled the operator without boundary conditions and use `E.setDeflationSpace` with the constant vector.

File: tests/regression/test_slepc.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    E.create()
+    E.setOperators(A, M)
+    st = E.getST()
+    st.setType('sinvert')
+    kspE = st.getKSP()
+    kspE.setType('fgmres')
+    pc = kspE.getPC()
+    pc.setType('lu')
+    if parallel:
+        pc.setFactorSolverPackage('mumps')
```
Comment (wence-) [inline tests/regression/test_slepc.py]: On travis we don't install petsc with support for mumps (since it takes rather a long time to build).  Is it possible to recast this in terms of an iterative solver (you're only actually looking at a few of the eigenmodes after all).
